test/e2e/cluster_deletion.go

```diff
@@ -0,0 +1,399 @@
+	"sigs.k8s.io/cluster-api/test/e2e/internal/log"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+)
+
+// ClusterDeletionSpecInput is the input for ClusterDeletionSpec.
+type ClusterDeletionSpecInput struct {
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+	SkipCleanup           bool
+
+	// Cluster name allows to specify a deterministic clusterName.
+	// If not set, a random one will be generated.
+	ClusterName *string
+
+	// InfrastructureProvider allows to specify the infrastructure provider to be used when looking for
+	// cluster templates.
+	// If not set, clusterctl will look at the infrastructure provider installed in the management cluster;
+	// if only one infrastructure provider exists, it will be used, otherwise the operation will fail if more than one exists.
+	InfrastructureProvider *string
+
+	// Flavor, if specified is the template flavor used to create the cluster for testing.
+	// If not specified, the default flavor for the selected infrastructure provider is used.
+	Flavor *string
+
+	// ControlPlaneMachineCount defines the number of control plane machines to be added to the workload cluster.
+	// If not specified, 1 will be used.
+	ControlPlaneMachineCount *int64
+
+	// WorkerMachineCount defines number of worker machines to be added to the workload cluster.
+	// If not specified, 1 will be used.
+	WorkerMachineCount *int64
+
+	// Allows to inject functions to be run while waiting for the control plane to be initialized,
+	// which unblocks CNI installation, and for the control plane machines to be ready (after CNI installation).
+	ControlPlaneWaiters clusterctl.ControlPlaneWaiters
+
+	// Allows to inject a function to be run after test namespace is created.
+	// If not specified, this is a no-op.
+	PostNamespaceCreated func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace string)
+
+	// Allows to inject a function to be run after machines are provisioned.
+	// If not specified, this is a no-op.
+	PostMachinesProvisioned func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace, workloadClusterName string)
+
+	// ClusterDeletionPhases define kinds which are expected to get deleted in phases.
+	ClusterDeletionPhases []ClusterDeletionPhase
+}
+
+type ClusterDeletionPhase struct {
+	// DeleteKinds are the DeleteKinds which are supposed to be deleted in this phase.
+	// Note: If Machines is added here, the Machine must have an owner with a kind from DeleteBlockKinds or DeleteKinds.
+	DeleteKinds sets.Set[string]
+	// DeleteBlockKinds are the kinds which are considered to be deleted in this phase, but need to be blocking.
+	// Note: If Machines is added here, the Machine must have an owner with a kind from DeleteBlockKinds or DeleteKinds.
```

Comment:
What would be an example case where the owner of the Machine is part of DeleteBlockKinds? (I only see it used with DeleteKinds)